README.md
#  1. Introducción
El siguiente proyecto constituye una SDK en PHP, que simplifica la ejecución de llamadas al servicio Ecommerce de Sipay.

# 2. Quickstart

Con el siguiente ejemplo podrás, en pocos pasos, instalar la SDK y efectuar una venta desde una terminal.
Antes de hacer el quickstart, debemos crear el archivo de configuraciones, tomando como referencia el indicado en la sección de configuración

```bash
  $ git clone https://github.com/sipay/php-sdk
  $ echo '
  <?php
  require_once "php-sdk/src/autoload.php";
  $ecommerce = new \Sipay\Ecommerce("php-sdk/etc/config.ini"); // Configurar el archivo de configuración como se indica en la sección Ecommerce
  $amount = new \Sipay\Amount(100, "EUR");
  $card = new \Sipay\Paymethods\Card("4242424242424242", 2050, 12);
  $options = array(
    "order" => "order-test",
    "reference" => "1234",
    "token" => "new-token"
  );
  $auth = $ecommerce->authorization($card, $amount, $options);
  if($auth->code == 0){
    print("Autorización aceptada, el pago ha sido completado!\n");
  }else{
    print("Error: ".$auth->description."\n");
  }
  ' > quickstart.php
  $ php quickstart.php
```
# 3. Instalación
## 3.1 Pre-requisitos
 PHP >= 7.x

## 3.2 Pasos
Podemos instalarlo de forma manual (Ver "Extracción de código") o utilizando Composer (Ver "Composer").

### 3.2.1 Extracción de código

Este paso se omite si se utiliza la instalación via composer.

  ```bash
    $ git clone https://github.com/sipay/php-sdk.git
  ```

Entonces establecemos la variable "path_autoload" a "<directorio_de_descarga_de_php-sdk>/src/autoload.php"

```php
$path_autoload = "<directorio_de_descarga_de_php-sdk>/src/autoload.php";
```

## 3.2.2 Composer

Este paso se omite si se utiliza el método manual.

```bash
$ composer require sipay/php-sdk

```

Entonces establecemos la variable "path_autoload" a "vendor/autoload.php":

```php
$path_autoload = "vendor/autoload.php";
```

# 4. Configuración
Una vez que se ha instalado la SDK, se deben actualizar los parámetros de configuración asociados a:
* Sistema de trazas.
* Credenciales de acceso (Se gestionan con el departamento de integraciones de Sipay).
* Entorno y versión de la API.
* Tiempo máximo de espera de respuestas (Timeout).
## 4.1. Desde un fichero

  Un ejemplo de configuraciones se muestra a continuación:
  ```ini
  ; **************************************************************
  ; LOGGER
  ; ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ; Configuración asociada al sistema de trazas.
  ;
  ; path: ruta del directorio de logs (Nota: Aconsejable usar rutas absolutas, en caso contrario los logs estaran dentro del paquete)
  ; level: nivel mínimo de trazas [debug, info, warning, error, critical]
  ; prefix: prefijo
  ; extension: extensión del archivo
  ; date_format: formato de fecha de las trazas
  ; backup_file_rotation: Número de ficheros de backup
  ; ------------------------------------------------------------//

  [logger]
  path=logs
  level=warning
  prefix=logger
  extension=log
  date_format=d/m/Y H:i:s
  backup_file_rotation = 5

  ; **************************************************************
  ; CREDENTIALS
  ; ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ; Credenciales para obtener acceso al recurso.
  ;
  ; key: Key del cliente
  ; secret: Secret del cliente
  ; resouce: Recurso al que se quiere acceder
  ; ------------------------------------------------------------//

  [credentials]
  key=api-key
  secret=api-secret
  resource=resource

  ; **************************************************************
  ; API
  ; ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ; Configuracion de la API.
  ;
  ; environment: Entorno al que se deben enviar las peticiones ['sandbox', 'staging', 'live']
  ; version: Versión de la api a usar actualmente solo existe v1
  ; mode: Modo de encriptacion de la firma, [sha256, sha512]
  ; ------------------------------------------------------------//

  [api]
  environment=sandbox
  version=v1
  mode=sha256

  ; **************************************************************
  ; Connection
  ; ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ; Cofiguracion de la conexión.
  ;
  ; timeout: Tiempo máximo de espera a la respuesta de la petición
  ; ------------------------------------------------------------//

  [connection]
  timeout=30
  ```

## 4.2. Desde una variable
En algunos casos, pudiera interesarnos cargar la configuración desde una variable y no desde un archivo. Este sería un ejemplo:

```php
  <?php
  $config = array(
      'logger' => array(
          'path' => 'logs',  // Nombre del directorio donde se crearan los logs (Nota: si la ruta es relativa se creará en el directorio del paquete)
          'level' => 'warning',  // Nivel mínimo de trazas [debug, info, warning, error, critical]
          'prefix' => 'logger',  // Prefijo del nombre del archivo
          'extension' => 'log',  // Extensión del archivo
          'date_format' => 'd/m/Y H:i:s',  // Formato de fecha de las trazas de log
          'backup_file_rotation' => 5  // Número de ficheros de backup
      ),
      'credentials' => array(
          'key' => 'api-key',  // Key del cliente
          'secret' => 'api-secret',  // Secret del cliente
          'resource' => 'resource' // Recurso al que se quiere acceder
      ),
      'api' => array(
          'environment' => 'sandbox',  // Entorno al que se deben enviar las peticiones ['sandbox', 'staging', 'live']
          'version' => 'v1',  // Versión de la api a usar actualmente solo existe v1
          'mode' => 'sha256'  // Modo de encriptacion de la firma, [sha256, sha512]
      ),
      'connection' => array(
          'timeout' => 30  // Tiempo máximo de respuesta de la petición.
      )
  );

```

# 5. Documentación extendida
A través de peticiones a Sipay mediante Ecommerce, se pueden realizar operativas de:
* Autorizaciones (sección 5.2.1).
* Cancelaciones (sección 5.2.2).
* Devoluciones (sección 5.2.3).
* Búsquedas de operaciones o querys (sección 5.2.4).
* Tokenización* de tarjetas (sección 5.2.5).
* Búsqueda de tarjetas tokenizadas (sección 5.2.6).
* Dar de baja una tarjeta tokenizada (sección 5.2.7).

_* Tokenización_: Es un proceso por el cual el PAN (_Primary Account Number_ – Número Primario de Cuenta) de la tarjeta se sustituye por un valor llamado token. Esta funcionalidad permite que Sipay guarde los datos de la tarjeta del cliente, para agilizar el proceso de pagos y evitar que se deba introducir, cada vez, los datos de tarjeta, en pagos repetitivos. Sipay realiza el almacenamiento de los datos de forma segura, cumpliendo con las normativas PCI.

Para llevar a cabo de forma correcta las operativas Ecommerce, se requiere el dominio de los objetos `Amount`, `Card`, `StoredCard` y `FastPay`, los cuales identifican el importe y el método de pago a utilizar.


## 5.1. Objetos necesarios en las operativas de Ecommerce

### **5.1.1. `Amount(amount,currency)`**

#### Definición
Este objeto representa una cantidad monetaria, por tanto esta cantidad debe ser mayor que cero (0). Para instanciar un objeto de este tipo se necesita una cantidad (amount) y una moneda (currency) en formato ISO4217 (https://en.wikipedia.org/wiki/ISO_4217).
La cantidad se puede especificar de tres formas:
* Con un `string` con la cantidad estandarizada y con el caracter punto (`.`) como separador de decimales (Ejemplo:  `"1.56"`), o
* Con un `string` que represente la cantidad en la unidad básica e indivisible de la moneda (Ejemplo: `"156"`), o
* Con un `int` que represente la cantidad en la unidad básica e indivisible de la moneda, es decir, la moneda Euro sería el céntimo (Ejemplo: `156`).


#### Parámetros
* **`amount`:**  [_obligatorio_] Es la cantidad de dinero a procesar. Se puede representar con un `string` o un `int`. Supongamos que queremos procesar 1.56 €, la cantidad (1.56) como un `string` sería `'1.56'` ó `'156'` ; como un `int` sería `156`.
* **`currency`:** [_obligatorio_] Es un `string` que representa el código de la moneda (ISO4217).

#### Atributos
* **`amount`:** `int` que representa la cantidad de procesamiento. Será este tipo de dato, independientemente de si se ha instanciado con un `string` previamente (`protected`).
* **`currency`:** `string` que representa el código de la moneda (ISO4217) (`protected`).

#### Métodos
* **`set(amount, currency)`:** Asigna un importe y moneda del tipo indicado en la sección de parámetros.
* **`get_amount()`:**  Devuelve el atributo amount.
* **`get_currency()`:** Devuelve el atributo currency.
* **`get_array()`:** Devuelve el amount y el currency como un array asociativo.

#### Ejemplo

```php
  <?php
  require_once $path_autoload;

  // Con string
  $amount = new \Sipay\Amount('1.56', 'EUR');

  print($amount."\n")
  // Imprime 1.56EUR
  print($amount->get_amount()."\n")
  // Imprime 156
  print($amount->get_currency()."\n")
  // Imprime EUR

  // Con unidad indivisible
  $amount = new \Sipay\Amount(156, 'EUR');

  print($amount."\n")
  // Imprime 1.56EUR
  print($amount->get_amount()."\n")
  // Imprime 156
  print($amount->get_currency()."\n")
  // Imprime EUR

  // Con string en unidad indivisible
  $amount = new \Sipay\Amount('156', 'EUR');

  print($amount."\n")
  // Imprime 1.56EUR
  print($amount->get_amount()."\n")
  // Imprime 156
  print($amount->get_currency()."\n")
  // Imprime EUR
```
**Nota:** En el caso de iniciarlo con un `string` que incluya un punto es imprescindible que tenga el número de decimales que indica el estándar ISO4217. Ejemplo: para la moneda euro (€) que establece dos decimales, es correcto indicar un amount de `"1.40"` pero no es correcto `"1.4"`.

### **5.1.2. `Card(card_number, year, month)`**

#### Definición
Este objeto representa una tarjeta que se puede utilizar en las diferentes operativas de Ecommerce. Para obtener una instancia de `Card`, los parámetros se indican a continuación.

#### Parámetros
* **`card_number`:** [_obligatorio_] Es un `string` con  longitud entre 14 y 19 dígitos. Representa el número de la tarjeta.
* **`year`:** [_obligatorio_] Es un `int` de 4 dígitos que indica el año de caducidad de la tarjeta.
* **`month`:** [_obligatorio_] Es un `int` de 2 dígitos con valores entre 1 y 12 que correspondiente al mes de caducidad de la tarjeta.

#### Atributos
* **`card_number`:** Es el número de la tarjeta en una instancia de `Card`. Es un `string` con longitud entre 14 y 19 dígitos (`protected`).
* **`year`:** Es al año de caducidad de la tarjeta en una instancia de `Card`. Es un  `int` de 4 dígitos (`protected`).
* **`month`:** Es el mes de caducidad de la tarjeta en una instancia de `Card`. Es un `int` de 2 dígitos entre 1 y 12 (`protected`).

#### Métodos
* **`set_card_number(card_number)`:** Permite asignar el PAN de la tarjeta
* **`set_expiration_date(year, month)`:** Permite asignar la fecha de caducidad de la tarjeta
* **`get_year()`:** Devuelve el atributo año.
* **`get_month()`:** Devuelve el atributo mes.
* **`get_card_number()`:** Devuelve el atributo card_number.
* **`to_json()`:** Devuelve todos los atributos del objeto.


#### Ejemplo
```php
  <?php
  require_once $path_autoload;

  $card = new \Sipay\Paymethods\Card('4242424242424242', 2018, 12);
```

### **5.1.3. `StoredCard(token)`**
#### Definición
Este objeto representa una tarjeta almacenada en Sipay que puede utilizarse en operativas Ecommerce. Para obtener una instancia de `StoredCard` se requieren los siguiente parámetros.

#### Parámetros
* **`token`:** [_obligatorio_] Es un `string` de longitud entre 6 y 128 caracteres de tipo alfanuméricos y guiones (`protected`).

#### Atributos
* **`token`:** `string` de longitud entre 6 y 128 caracteres.

#### Métoddos
* **`set_token`:**  Permite asignar el token.
* **`get_token`:**  Devuelve el token.

#### Ejemplo
```php
  <?php
  require_once $path_autoload;

  $card = new \Sipay\Paymethods\StoredCard('token-card');
```

### **5.1.4. `FastPay(token)`**

#### Definición
Este objeto representa una tarjeta obtenida mediante el método de pago FastPay. Se utiliza en los consecutivos pasos de la operativas de pago de este método.

#### Parámetros
* **`token`:** [_obligatorio_] Es un `string` de longitud 32 con caracteres de tipo hexadecimal.

#### Atributos
* **`token`:**`string` de longitud 32 caracteres de tipo hexadecimal (`protected`).

#### Métoddos
* **`set_token`:**  Permite asignar el token.
* **`get_token`:**  Devuelve el token.

#### Ejemplo
Ejemplo:
```php
  <?php
  require_once $path_autoload;

  $card = new \Sipay\Paymethods\FastPay('token-fast-pay');
```

## 5.2. Operativas de Ecommerce - `Ecommerce($config_file)` ó `Ecommerce($config_var)`

#### Descripción
Las operativas de Ecommerce forman parte de los métodos definidos en la clase `Ecommerce`. Para instanciar un objeto de este tipo se requiere el archivo de configuración.

#### Ejemplo
```php
  <?php
  require_once $path_autoload;

  $ecommerce = \Sipay\Ecommerce('etc/config.ini');
```

#### Parámetros
* **`config_file`** [_obligatorio_] Es un `string` con la ruta del archivo de configuraciones.

#### Atributos
Los siguientes atributos (`protected`) se asignan en el archivo de configuraciones. Sin embargo, son accesibles en instancias de `Ecommerce`. Se sugiere que sean utilizados en modo de consulta.
* **`key`:** corresponde al key de las credenciales.
* **`secret`:** corresponde al secret de las credenciales.
* **`resource`:** corresponde al resource de las credenciales.
* **`environment`:** corresponde al entorno al cual se está apuntando.
* **`mode`:** corresponde el modo de cifrado de las peticiones.
* **`version`:** corresponde a la versión de la API a la cual se apunta.
* **`timeout`:** Corresponde al tiempo de espera máximo en recibir respuesta contando desde el momento en que se envía la petición.

#### Métodos
Todos los atributos indicados tienen sus métodos de asignación con `set_[nombre_del_atributo]` y sus métodos de consulta con `get_[nombre_del_atributo]`.
* **`authorization(parameters)`:** Permite hacer peticiones de autorización haciendo uso de los diferentes métodos de pago (ver sección 5.2.1).
* **`cancellation(parameters)`:** Permite enviar peticiones de cancelaciones (ver sección 5.2.2).
* **`refund(parameters)`:** Permite hacer devoluciones (ver sección 5.2.3).
* **`query(parameters)`:** Permite hacer peticiones de búsqueda de operaciones (ver sección 5.2.4).
* **`register(parameters)`:** Permite tokenizar una tarjeta (ver sección 5.2.5).
* **`card(parameters)`:** Se utiliza para buscar una tarjeta que fue tokenizada (ver sección 5.2.6).
* **`unregister(parameters)`:** Se utiliza para dar de baja una tarjeta tokenizada (ver sección 5.2.7).

## 5.2.1 **`authorization(Paymethods\Paymethod $paymethod, Amount $amount, array $array_options = array())`**

### Definición
 Este método de `Ecommerce` permite enviar una petición de venta a Sipay.
### Parámetros
* **`paymethod`:**[_obligatorio_] Corresponde a una instancia  `Card`, `StoredCard` o `FastPay` que indica el método de pago a utilizar.
* **`amount `:** [_obligatorio_] Corresponde a una instancia de `Amount` que representa el importe de la operación.
* **`array_options `:**
  * **`order `:** [_opcional_] Es un `string` que representa el ticket de la operación.
  * **`reconciliation `:** [_opcional_] Es un `string` que identifica la conciliación bancaria.
  * **`custom_01` :** [_opcional_] Es un `string` que representa un campo personalizable.
  * **`custom_02` :** [_opcional_] Es un `string` que representa un campo personalizable.
  * **`token`:** [_opcional_] Es un `string` que representa un token a almacenar. Se utiliza cuando el método de pago es de tipo `Card` o `Fpay`, y se desea asignar un token específico a la tarjeta utilizada.

### Salida
El método `authorization` devuelve un objeto `Authorization`.

### Ejemplo
**- Autorización con tarjeta**
```php
  $amount = \Sipay\Amount(100, 'EUR'); // 1€
  $card = \Sipay\Paymethods\Card('4242424242424242', 2050, 2);

  $auth = $ecommerce->authorization($card, $amount);
```
**- Autorización con FastPay**
```php
   $amount = \Sipay\Amount(100, 'EUR'); // 1€
   $fp = \Sipay\Paymethods\FastPay('830dc0b45f8945fab229000347646ca5');

   $auth = $ecommerce->authorization($fp, $amount);
```

## 5.2.2 `cancellation(string $transaction_id)`

### Definición
Este método permite enviar una petición de cancelación a Sipay

### Parámetros
* **`transaction_id`:** [_obligatorio_] Es un `string` con el identificador de la transacción.

### Salida
El método `cancellation` devuelve un objeto `Cancellation`.

### Ejemplo
```php
  $cancel = $ecommerce->cancellation('transaction_id');
```
## 5.2.3 `refund($identificator, Amount $amount, array $array_options = array())`


### Definición
Este método `Ecommerce` permite enviar una petición de devolución a Sipay.
### Parámetros
* **`identificator`:** [_obligatorio_] Es una instancia del método de pago (`Card`, `StoredCard` o `FastPay`) o, un `string` que representa el identificador de transacción.
* **`amount `:** [_obligatorio_] Corresponde a una instancia de `Amount` que representa el importe de la operación.
* **`array_options `:**
  * **`order `:** [_opcional_] Es un `string` que representa el ticket de la operación.
  * **`reconciliation `:** [_opcional_] Es un `string` que identifica la conciliación bancaria.
  * **`custom_01` :** [_opcional_] Es un `string` que representa un campo personalizable.
  * **`custom_02` :** [_opcional_] Es un `string` que representa un campo personalizable.
  * **`token`:** [_opcional_] Es un `string` que representa un token a almacenar. Se utiliza cuando el método de pago es de tipo `Card` o `Fpay`, y se desea asignar un token específico a la tarjeta utilizada.


### Salida
El método `refund` devuelve un objeto `Refund`.

### Ejemplo
**- Devolución con tarjeta**
```php
  $amount = \Sipay\Amount(100, 'EUR'); // 1€
  $card = \Sipay\Paymethods\StoredCard('bd6613acc6bd4ac7b6aa96fb92b2572a');
  $refund = $ecommerce->refund($card, $amount);
```

**- Devolución con transaction_id**
```php
  $amount = \Sipay\Amount(100, 'EUR'); // 1€

  $refund = $ecommerce->refund('transaction_id', $amount);
```

## 5.2.5 `query(array $query)`

### Definición
Este método `Ecommerce` permite enviar una petición a Sipay para buscar de una operación concreta.

### Parámetros
 El método puede tener los siguientes parámetros:
* **`order`:** [_opcional_] `string` que representa el ticket de la operación.
* **`transaction_id`:** [_opcional_]  `string` que representa el identificador de la transacción.

En caso de no enviar ninguno, se devuelven todas las transacciones del cliente.

### Salida
El método `query` devuelve un objeto `Query`.

### Ejemplo
**- Búsqueda de transacciones**

```php
  $query = array('transaction_id' => 'transaction_id');
  $query = $ecommerce->query($query);
```

## 5.2.6 `register(Paymethods\Paymethod $card, string $token)`

### Definición
Este método `Ecommerce` permite enviar una petición de tokenización de tarjeta a Sipay.

### Parámetros
  * **`card`:** [_obligatorio_] Instancia de tipo `Card` con la tarjeta a tokenizar o tipo `FastPay`.
  * **`token`:**[_obligatorio_] `string` que representa el token que se asociará a la tarjeta.

### Salida
    El método `register` devuelve un objeto `Register`.

###  Ejemplo

  **- Registro de tarjeta**
  ```php
    $card = \Sipay\Paymethods\Card('4242424242424242', 2050, 2);

    $masked_card = $ecommerce->register($card, 'newtoken');
  ```

## 5.2.7 `card(string $token)`

### Definición
Este método `Ecommerce` permite enviar una petición a Sipay con la finalidad de obtener información de una tarjeta que está tokenizada.

### Parámetros
* **`token`:**[_obligatorio_] `string` que representa el token asociado a la tarjeta.

### Salida
El método `card` devuelve un objeto `Card` del apartado Responses.

###  Ejemplo
**- Búsqueda de tarjeta**

```php
   $masked_card = $ecommerce->card('newtoken');
```

## 5.2.8 `unregister(string $token)`

### Definición
Este método `Ecommerce` permite enviar una petición a Sipay con la finalidad de dar de baja una tarjeta tokenizada.

### Parámetros
* **`token`:**[_obligatorio_] `string` que representa el token asociado a la tarjeta.

### Salida
El método `unregister` devuelve un objeto `Unregister`.

###  Ejemplo
**- Borrar una tarjeta del registro**
```php
  $unregister = $ecommerce->unregister('newtoken');
```

### 5.3 Responses
Todos los objetos obtenidos como respuestas de operativas `Ecommerce` tienen los siguientes atributos.

#### 5.3.1 Atributos comunes

* **`type`:** Es un `enum[string]` que identifica el tipo de respuesta:
    * success
    * warning
    * error
* **`code`:** Es un `int` con el código identificador del resultado. Es un código orientativo y no está ligado estrictamente con motivo de la respuesta, es decir, el código no identifica unívocamente la respuesta.
    - si `code` es `0`, implica que el resultado es un _success_
    - si `code` es mayor a `0`, implica que el resultado es un _warning_
    - si `code` es menor a `0`, implica que el resultado es un _error_
* **`detail`:**  Es un `string` con el código alfanumérico separado con guiones bajos y sin mayúsculas que identifica unívocamente la respuesta. Útil para la gestión de los diferentes casos de uso de una operación.
* **`description`:** Es un `string` con la descripción literal del mensaje de respuesta.
* **`uuid`:** Es un `string` con el identificador único de la petición, imprescindible para la trazabilidad.
* **`request_id`:** Es un `string` utilizado en la finalización de algunas operaciones. Se indicarán aquellas en las que sea necesario.
* **`request`:** Es un `dictionary` que contiene los datos de la petición que se ha hecho al servidor (`protected`).
* **`response`:** Es un `dictionary` que contiene los datos 'raw' de respuesta (`protected`).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased] - XXXX-XX-XX

## [1.0.3] - 2018-05-22
### Changed
- Acualizada la longitud del key value a 128b
## [1.0.2] - 2018-03-08
### Added
- Añadiendo configuracion por variable
## [1.0.1] - 2018-03-08
### Added
- Propietario en el archivo composer

## [1.0.0] - 2018-03-08
### Added
- Implementación incial de la SDK con operativas:
	- Autorizacion mediante diferentes métodos(tarjeta, token, fastpay)
	- Cancelación
	- Devolución
	- Query
	- Tokenización
- Intrucciones de instalación y ejecución
- Ejemplos para el uso de la SDK


[Unreleased]: https://github.com/Sipay/php-sdk/compare/1.0.3...develop
[1.0.3]: https://github.com/Sipay/php-sdk/compare/1.0.2...1.0.3
[1.0.2]: https://github.com/Sipay/php-sdk/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/Sipay/php-sdk/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/Sipay/php-sdk/compare/8618da1298a4f8cbe0c3a5fb7cc31cb642d4677f...1.0.0




src/Ecommerce.php
<?php

namespace Sipay;

class Ecommerce
{
    protected $logger;
    protected $key;
    protected $secret;
    protected $resource;
    protected $environment;
    protected $version;
    protected $mode;
    protected $timeout;

    public function __construct($config)
    {
        if (gettype($config) == "string") {

          $config_obj = parse_ini_file($config, true);

        }else if(gettype($config) == "array"){
          $config_obj = $config;
        }else{
          throw new \Exception('$config must be a string or array.');
        }

        $this->logger = new Logger($config_obj['logger']);

        $cred = $config_obj['credentials'];
        $this->setKey(isset($cred['key']) ? $cred['key'] : '');
        $this->setSecret(isset($cred['secret']) ? $cred['secret'] : '');
        $this->setResource(isset($cred['resource']) ? $cred['resource'] : '');

        $api = $config_obj['api'];
        $this->setEnvironment(isset($api['environment']) ? $api['environment'] : '');
        $this->setVersion(isset($api['version']) ? $api['version'] : '');
        $this->setMode(isset($api['mode']) ? $api['mode'] : '');

        $connection = $config_obj['connection'];
        $this->setTimeout(isset($connection['timeout']) ? $connection['timeout'] : '30');
    }

    public function getKey()
    {
        return $this->key;
    }

    public function setKey($key)
    {
        if (gettype($key) != "string") {
            throw new \Exception('$key must be a string.');
        }
        if(!preg_match("/^[\w-]{6,32}$/", $key)) {
            throw new \Exception('Invalid key provided: wrong format.');
        }
        $this->key = $key;
    }

    public function getSecret()
    {
        return $this->secret;
    }

    public function setSecret($secret)
    {
        if (gettype($secret) != "string") {
            throw new \Exception('$secret must be a string.');
        }
        if(!preg_match("/^[\w-]{6,128}$/", $secret)) {
            throw new \Exception('Invalid secret provided: wrong format.');
        }
        $this->secret = $secret;
    }

    public function getResource()
    {
        return $this->resource;
    }

    public function setResource($resource)
    {
        if (gettype($resource) != "string") {
            throw new \Exception('$resource must be a string.');
        }
        if(!preg_match("/^[\w-]{6,32}$/", $resource)) {
            throw new \Exception('Invalid resource provided: wrong format.');
        }
        $this->resource = $resource;
    }

    public function getEnvironment()
    {
        return $this->environment;
    }

    public function setEnvironment($environment)
    {
        if (gettype($environment) != "string") {
            throw new \Exception('$environment must be a string.');
        }
        if(!in_array($environment, array('sandbox', 'staging', 'live'))) {
            throw new \Exception('$environment must be sandbox, staging or live.');
        }
        $this->environment = $environment;
    }

    public function getVersion()
    {
        return $this->version;
    }

    public function setVersion($version)
    {
        if (gettype($version) != "string") {
            throw new \Exception('$version must be a string.');
        }
        if($version != "v1") {
            throw new \Exception('$version must be v1.');
        }
        $this->version = $version;
    }

    public function getMode()
    {
        return $this->mode;
    }

    public function setMode($mode)
    {
        if (gettype($mode) != "string") {
            throw new \Exception('$mode must be a string.');
        }
        if(!in_array($mode, array('sha256','sha512'))) {
            throw new \Exception('$mode must be sha256 or sha512.');
        }
        $this->mode = $mode;
    }

    public function getTimeout()
    {
        return $this->timeout;
    }

    public function setTimeout($timeout)
    {
        if (!is_numeric($timeout)) {
            throw new \Exception('$timeout must be a integer.');
        }
        $timeout = intval($timeout);
        if($timeout <= 0) {
            throw new \Exception('$timeout must be greater than 0.');
        }
        $this->timeout = $timeout;
    }

    private function send($payload, $endpoint)
    {
        $this->logger->info(
            'sipay.request', 'request.send', 'I-0001', 'Request Send',
            array('payload' => $payload, 'endpoint' => $endpoint)
        );

        $url = 'https://'.$this->environment.'.sipay.es/mdwr/'.$this->version.'/'.$endpoint;
        $data = array(
            'key' => $this->key,
            'nonce' => "".time(),
            'mode' => $this->mode,
            'resource' => $this->resource,
            'payload' => $payload
        );

        $body = json_encode($data);
        $signature = hash_hmac($this->mode, $body, $this->secret);

        $options = array(
            'http' => array(
                'header'  => "Content-type: application/json\r\nContent-Signature: ".$signature."\r\n",
                'method'  => 'POST',
                'content' => $body,
                'timeout' => $this->timeout,
                'ignore_errors' => true
            ),
        );

        $context = stream_context_create($options);
        try {
            $response_body = file_get_contents($url, false, $context);
        } catch (Exception $e) {
            $this->logger->error(
                'sipay.request', 'request.response', 'E-0004', 'Request Error',
                array('error_msg' => $e->getMessage())
            );
            $response = null;
            $response_body = null;
        }

        if(!is_null($response_body)) {
            if(isset($http_response_header)) {
                if($response_body == "") {
                    $this->logger->error('sipay.request', 'request.response', 'E-0001', 'Request Error', array());
                    $response = null;

                }else{
                    $is_json = false;
                    foreach($http_response_header as $header){
                        if(substr($header, 0, 30) === "Content-Type: application/json") {
                            $is_json = true;
                            break;
                        }
                    }

                    if($is_json) {
                        $response = json_decode($response_body, true);

                    }else{
                        $this->logger->error('sipay.request', 'request.response', 'E-0003', 'Response no json', array());
                        $response = null;
                    }
                }

            }else{
                $this->logger->error('sipay.request', 'request.response', 'E-0002', 'Request Error', array());
                $response = null;
            }
        }

        $this->logger->info(
            'sipay.request', 'request.response', 'I-0002', 'Request Response',
            array(
               'payload' => $payload,
               'endpoint' => $endpoint,
               'response' => $response_body
            )
        );

        return array($body, $response);
    }

    private function clean_parameters(array $array_options, array $array_schema)
    {
        $options = array();
        foreach ($array_options as $name => $option) {
            if(isset($array_schema[$name])) {
                $schema = $array_schema[$name];

                if(gettype($option) != $schema['type']) {
                    throw new \Exception("$name incorrect type.");
                }

                if(is_string($option) && isset($schema['pattern']) && !preg_match($schema['pattern'], $option)) {
                    throw new \Exception('Invalid name provided: wrong format.');
                }

                $options[$name] = $option;
            }
        }

        return $options;

    }

    public function authorization(Paymethods\Paymethod $paymethod, Amount $amount, array $array_options = array())
    {

        $array_schema = array(
            'order' => array(
                'type' => 'string',
                'pattern' => '/^[\w-]{6,64}$/'
            ),
            'reference' => array(
                'type' => 'string',
                'pattern' => '/^[0-9]{4}[a-zA-Z0-9]{0,8}$/'
            ),
            'token' => array(
                'type' => 'string',
                'pattern' => '/^[\w-]{6,128}$/'
            ),
            'custom_01' => array(
                'type' => 'string'
            ),
            'custom_02' => array(
                'type' => 'string'
            ),
          );

        $options = $this->clean_parameters($array_options, $array_schema);

        if(isset($options['reference'])) {
            $options['reconciliation'] = $options['reference'];
            unset($options['reference']);
        }


        $payload = array_replace($options, $amount->get_array(), $paymethod->to_json());

        $args = $this->send($payload, 'authorization');
        return new Responses\Authorization($args[0], $args[1]);

    }

    public function refund($identificator, Amount $amount, array $array_options = array())
    {
        $is_paymethod = is_subclass_of($identificator, 'Sipay\Paymethods\Paymethod');
        $is_tx_id = gettype($identificator) == "string" && preg_match('/^[0-9]{6,22}$/', $identificator);

        if (!$is_paymethod && !$is_tx_id) {
            throw new \Exception('Incorrect $identificator.');
        }

        $array_schema = array(
            'order' => array(
                'type' => 'string',
                'pattern' => '/^[\w-]{6,64}$/'
            ),
            'reference' => array(
                'type' => 'string',
                'pattern' => '/^[0-9]{4}[a-zA-Z0-9]{0,8}$/'
            ),
            'token' => array(
                'type' => 'string',
                'pattern' => '/^[\w-]{6,128}$/'
            ),
            'custom_01' => array(
                'type' => 'string'
            ),
            'custom_02' => array(
                'type' => 'string'
            ),
          );

        $options = $this->clean_parameters($array_options, $array_schema);

        if(isset($options['reference'])) {
            $options['reconciliation'] = $options['reference'];
            unset($options['reference']);
        }

        if($is_paymethod) {
            $id_array = $identificator->to_json();
        }else{
            $id_array = array('transaction_id' => $identificator);
        }

        $payload = array_replace($options, $amount->get_array(), $id_array);

        $args = $this->send($payload, 'refund');
        return new Responses\Refund($args[0], $args[1]);

    }

    public function register(Paymethods\Paymethod $card, string $token)
    {

        if ($card instanceof Paymethods\StoredCard) {
            throw new \Exception('$card can\'t be a StoredCard instance.');
        }

        if(!preg_match('/^[\w-]{6,128}$/', $token)) {
            throw new \Exception('Invalid token provided: wrong format.');
        }

        $payload = array(
          'token' => $token
        );

        $payload = array_replace($payload, $card->to_json());

        $args = $this->send($payload, 'register');
        return new Responses\Register($args[0], $args[1]);

    }

    public function card(string $token)
    {
        if(!preg_match('/^[\w-]{6,128}$/', $token)) {
            throw new \Exception('Invalid token provided: wrong format.');
        }

        $payload = array(
          'token' => $token
        );

        $args = $this->send($payload, 'card');
        return new Responses\Card($args[0], $args[1]);

    }

    public function unregister(string $token)
    {
        if(!preg_match('/^[\w-]{6,128}$/', $token)) {
            throw new \Exception('Invalid token provided: wrong format.');
        }

        $payload = array(
          'token' => $token
        );

        $args = $this->send($payload, 'unregister');
        return new Responses\Unregister($args[0], $args[1]);

    }

    public function cancellation(string $transaction_id)
    {

        if(!preg_match('/^[0-9]{6,22}$/', $transaction_id)) {
            throw new \Exception('Invalid transaction_id provided: wrong format.');
        }

        $payload = array(
          'transaction_id' => $transaction_id
        );

        $args = $this->send($payload, 'cancellation');
        return new Responses\Cancellation($args[0], $args[1]);

    }

    public function query(array $query)
    {
        $payload = array();

        $array_schema = array(
            'order' => array(
                'type' => 'string',
                'pattern' => '/^[\w-]{6,64}$/'
            ),
            'transaction_id' => array(
                'type' => 'string',
                'pattern' => '/^[0-9]{6,22}$/'
            )
          );

        $payload = $this->clean_parameters($query, $array_schema);

        $args = $this->send($payload, 'query');
        return new Responses\Query($args[0], $args[1]);
    }
}
